[Basic] Add `RerootedFileSystemView`.

ddunbar opened the pull request:
- This is a wrapper object which presents a view on an existing FileSystem as
   rooted at a subpath.

ddunbar commented:
@swift-ci please test

/cc @abertelrud I'm not in love with the name, but I couldn't come up with an alternative I liked better.

ddunbar commented:
... I'm also trying to find a nicer name for the initializer.

abertelrud commented:
Cool, I like the concept!  Name:  `FileSystemSlice`?  In spirit, that would be in keeping with the Swift philosophy.

abertelrud commented:
Except that in a collection slice, the indices aren't shifted.  Whereas here, you're shifting the indices (paths) inside the slice, right?  So maybe that's not a great name.

ddunbar commented:
Interesting idea, I could go with that. Does it convey the path-based semantics enough or does it sound like any arbitrary slice on top an existing filesystem?

abertelrud commented:
Right, that's what I was thinking in my second comment:  a true file system slice, in the sense of Swift collections, would be just kind of a sandboxing, where the paths are the same but only a subset of them are accessible.  This is more rebased, or rerooted.  Maybe just `RerootedFileSystem`?  Or `SubFileSystem`?

ddunbar commented:
I like `RerootedFileSystem` or `SubpathFileSystem` (but prefer the former).

What about `FileSystemSubpathView`?

ddunbar commented:
`...View` is also useful here to connote that mutations to it are going to mutate the input.

abertelrud commented:
Yeah, that's a good one.  `FileSystemSubpathView` or `RerootedFileSystemView` maybe.  And clients still see just a `let fileSys: FileSystem` in terms of their API, right?

abertelrud commented:
(clients that take a file system as input and access it, that is — of course the client that creates the subpath view has to mention it by name)

ddunbar commented:
Yeah... the goal is that most clients will just take a `FileSystem` and not care about where it comes from. The expectation is then that we can get a `FileSystem` view on a bare git repo, or a `RerootedFileSystemView` on an actual file system, and the client doesn't care.

abertelrud commented:
Perfect.

ddunbar commented:
Renamed to `RerootedFileSystem` and adjusted the initializer to a name I prefer better.

I'll land this at some point if I don't hear other objections.

abertelrud commented on Sources/Basic/FileSystem.swift:
```diff
@@ -467,5 +466,60 @@ public class InMemoryFileSystem: FileSystem {
+/// of an existing filesystem. This is useful for passing to clients which only
+/// need access to a subtree of the filesystem but should otherwise remain
+/// oblivious to its concrete location.
+public struct RerootedFileSystemView: FileSystem {
```
Comment is great, but perhaps also add some clarification that this doesn't provide any type of security, e.g. a symlink is allowed to escape the subtree.

ddunbar replied:
Agreed, will do.

ddunbar replied:
Done here: 9653434

abertelrud commented on Sources/Basic/FileSystem.swift:
```diff
@@ -467,5 +466,60 @@ public class InMemoryFileSystem: FileSystem {
+            return root
+        } else {
+            // FIXME: Optimize?
+            return root.appending(RelativePath(String(path.asString.characters.dropFirst(1))))
```
I have actually long been of the opinion that a leading path separator should be allowed when the caller is specifically appending a subpath to an existing path, as is done in rerooting cases (e.g. a build system that appends `$(INSTALL_DIR)` to `$(INSTALL_ROOT)`).  I'm fine if the default behavior is more strict, but there should be some way, I think, to allow inputs to be a bit more flexible.

ddunbar replied:
I agree, it would be nice to have a method for this.

abertelrud commented on Sources/Basic/FileSystem.swift:
```diff
@@ -467,5 +466,60 @@ public class InMemoryFileSystem: FileSystem {
+    /// The root path within the containing file system.
+    private let root: AbsolutePath
+    
+    public init(_ underlyingFileSystem: inout FileSystem, rootedAt root: AbsolutePath) {
```
Hmm, why is that `inout`?  That struct doesn't get changed, right?

ddunbar replied:
It does in fact get changed by the mutating methods (the test case actually uses this).

It would be nice if it could be `inout` only in cases where the client wants a `var`, but the type system can't do this.

ddunbar replied:
... this means that a client which got a FileSystem, but wants to expose an (readonly) `FileSystem` to a client, will still have to make a copy to create the rerooted FileSystem (even though the copy won't be modified). That's unfortunate, but at least is localized.

abertelrud commented:
A couple of comments, but apart from that, it looks like a nice improvement.

ddunbar commented:
Manually tested on macOS and Linux.

ddunbar commented:
Closing, I accidentally landed by rebasing onto master.
